homeassistant/components/mikrotik/__init__.py

```diff
@@ -1 +1,390 @@
+            login_method = (login_token,)
+        else:
+            login_method = (login_plain, login_token)
+
+        kwargs = {
+            'encoding': config[CONF_ENCODING],
+            'login_methods': login_method
+        }
+
+        if CONF_PORT in config:
+            kwargs['port'] = config[CONF_PORT]
+        else:
+            kwargs['port'] = MTK_DEFAULT_API_PORT
+
+        if config[CONF_SSL]:
+            if CONF_PORT not in config:
+                kwargs['port'] = MTK_DEFAULT_API_SSL_PORT
+            ssl_context = ssl.create_default_context()
+            ssl_context.check_hostname = False
+            ssl_context.verify_mode = ssl.CERT_NONE
+            kwargs['ssl_wrapper'] = ssl_context.wrap_socket
+        host = config.get(CONF_HOST)
+        self._hosts[host]['kwargs'] = kwargs
+
+    def get_polling_method(self, host):
+        """Get Mikrotik device_tracker polling method."""
+        import librouteros
+        polling_method = self._hosts[host]['config'].get(CONF_METHOD)
+        try:
+            capsman_exist = self._client[host](
+                cmd=MIKROTIK_SERVICES[CAPSMAN])
+        except (librouteros.exceptions.TrapError,
+                librouteros.exceptions.MultiTrapError,
+                librouteros.exceptions.ConnectionError):
+            _LOGGER.info("Mikrotik %s: Not a CAPsMAN controller. Trying "
+                         "local wireless interfaces.", (host))
+            capsman_exist = False
+        try:
+            wireless_exist = self._client[host](
+                cmd=MIKROTIK_SERVICES[WIRELESS])
+        except (librouteros.exceptions.TrapError,
+                librouteros.exceptions.MultiTrapError,
+                librouteros.exceptions.ConnectionError):
+            wireless_exist = False
+
+        if not wireless_exist and not capsman_exist \
+                or polling_method == 'ip':
+            _LOGGER.info(
+                "Mikrotik %s: Wireless adapters not found. Try to "
+                "use DHCP lease table as presence tracker source. "
+                "Please decrease lease time as much as possible",
+                host)
+        if polling_method:
+            _LOGGER.info("Mikrotik %s: Manually selected polling method %s",
+                         host, polling_method)
+            device_tracker = polling_method
+        else:
+            if capsman_exist:
+                device_tracker = CAPSMAN
+            elif wireless_exist:
+                device_tracker = WIRELESS
+            else:
+                device_tracker = DHCP
+            _LOGGER.info("Mikrotik %s: Using device_tracker method %s",
+                         host, device_tracker)
+        self._hosts[host][DEVICE_TRACKER] = device_tracker
+
+    def connect_to_device(self, host):
+        """Connect to Mikrotik method."""
+        import librouteros
+        import time
+        tick = 0
+        while self._hosts[host][CONNECTING] and tick < 60:
+            time.sleep(1)
```

Comment:
We're not allowed to sync sleep in the event loop. We're in async context, ie the event loop here. It's ok to use `asyncio.sleep`, but then we need to be inside a coroutine to be able to await.

Why do we need to sleep?

Reply:
When I was testing an MikrotikAPI disconnect from the device there could be more than one call to connect_to_device(). What you explained makes sense and if another call is connecting it should just return. Fixing.